### PR TITLE
First Version of Peer manager

### PR DESCRIPTION
**Pull request by ckoopmann**
I went ahead, dabbled a bit and added a first version of a "Peer Manager" that tries to maintain a certain minimum number of connected peers.

The Peer Manager contains the following steps:

1. Maintain a "peer_data" mapping that, keeps track of peers "Performance" (currently simply measured by the time that they stayed connected before disconnecting us in the past) and their connection history. Includes persistiing this data between runs in a json file
2. Regularly check if additional peers need to be dialled (by comparing connected number of peers and target number)
3. Find peers to redial (based on past performance)
4. Ask Discovery Service to find / discover new peers if necessary 
5. Add newly discovered Peers coming from the Discovery Service
6. Keep track of peers "identity" as reported by the "identify" service. (Although this data is not used anywhere yet)

This implementation does not yet contain any logic to down-regulate the peer count (i.e. disconnect peers) if it exceeds the target

### CODE REVIEW

File: src/config.rs [hunk 1]
```diff
@@ -1,7 +1,15 @@
...
+	"enr:-Ku4QG-2_Md3sZIAUebGYT6g0SMskIml77l6yR-M_JXc-UdNHCmHQeOiMLbylPejyJsdAPsTHJyjJB2sYGDLe0dn8uYBh2F0dG5ldHOIAAAAAAAAAACEZXRoMpC1MD8qAAAAAP__________gmlkgnY0gmlwhBLY-NyJc2VjcDI1NmsxoQORcM6e19T1T9gi7jxEZjk_sjVLGFscUNqAY9obgZaxbIN1ZHCCIyg",
+	"enr:-Ku4QPn5eVhcoF1opaFEvg1b6JNFD2rqVkHQ8HApOKK61OIcIXD127bKWgAtbwI7pnxx6cDyk_nI88TrZKQaGMZj0q0Bh2F0dG5ldHOIAAAAAAAAAACEZXRoMpC1MD8qAAAAAP__________gmlkgnY0gmlwhDayLMaJc2VjcDI1NmsxoQK2sBOLGcUb4AwuYzFuAVCaNHA-dy24UuEKkeFNgCVCsIN1ZHCCIyg",
+	"enr:-Ku4QEWzdnVtXc2Q0ZVigfCGggOVB2Vc1ZCPEc6j21NIFLODSJbvNaef1g4PxhPwl_3kax86YPheFUSLXPRs98vvYsoBh2F0dG5ldHOIAAAAAAAAAACEZXRoMpC1MD8qAAAAAP__________gmlkgnY0gmlwhDZBrP2Jc2VjcDI1NmsxoQM6jr8Rb1ktLEsVcKAPa08wCsKUmvoQ8khiOl_SLozf9IN1ZHCCIyg",
+];
+
 // One of the Ethereum Foundation Geth Bootnodes
-pub const BOOTNODE: &str = "enr:-Ku4QG-2_Md3sZIAUebGYT6g0SMskIml77l6yR-M_JXc-UdNHCmHQeOiMLbylPejyJsdAPsTHJyjJB2sYGDLe0dn8uYBh2F0dG5ldHOIAAAAAAAAAACEZXRoMpC1MD8qAAAAAP__________gmlkgnY0gmlwhBLY-NyJc2VjcDI1NmsxoQORcM6e19T1T9gi7jxEZjk_sjVLGFscUNqAY9obgZaxbIN1ZHCCIyg";
+pub const BOOTNODE: &str = BOOTNODES[0];
```
Comment (brech1) [inline src/config.rs]: This unused const can be removed

File: src/discovery.rs [hunk 2]
```diff
@@ -9,6 +9,7 @@ use libp2p::identity::Keypair;
 use libp2p::multiaddr::Protocol;
 use libp2p::swarm::{NetworkBehaviour, NetworkBehaviourAction, PollParameters};
 use libp2p::{Multiaddr, PeerId};
+use log::{debug, error, info, trace, warn};
```
Comment (brech1) [inline src/discovery.rs]: Some of the log instances are not used! They should be removed

File: src/main.rs [hunk 3]
```diff
@@ -21,24 +23,33 @@ mod chain;
 mod config;
 mod discovery;
 mod enr;
+mod peer_manager;
 mod rpc;
 use crate::rpc::methods::{
     EnrAttestationBitfield, EnrSyncCommitteeBitfield, MetaData, RPCCodedResponse, RPCResponse,
 };
+use std::collections::{HashMap, HashSet};
```
Comment (brech1) [inline src/main.rs]: These collections are not being used

File: src/main.rs [hunk 4]
```diff
@@ -176,16 +190,50 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
...
                         rpc::RPCReceived::Response(_, _) => todo!(),
                         },
-                        Err(e) =>  println!("{}", e),
+                        Err(e) =>  warn!("Error in RPC quest handling: {}", e),
                     }},
-                    BehaviourEvent::Identify(ev) => println!("identify: {:#?}", ev),
+                    BehaviourEvent::Identify(ev) => {
+                        debug!("identify: {:#?}", ev);
+                        match ev {
+                            identify::IdentifyEvent::Received { peer_id, info, .. } => {
```
Comment (brech1) [inline src/main.rs]: `IdentifyEvent` seems to be deprecated, but it can be replaced by `libp2p::identify::Event`

File: src/main.rs [hunk 5]
```diff
@@ -123,36 +136,37 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
         .connection_event_buffer_size(64)
         .connection_limits(
             ConnectionLimits::default()
-                .with_max_pending_incoming(Some(5))
-                .with_max_pending_outgoing(Some(16))
-                .with_max_established_per_peer(Some(1)),
+                .with_max_pending_incoming(Some(64))
+                .with_max_pending_outgoing(Some(32))
+                .with_max_established_per_peer(Some(10)),
```
Comment (brech1) [inline src/main.rs:140]: You enabled more than one concurrent connection per peer, am I correct? Have you noticed any improvements?

**Review by brech1 — COMMENTED** [review]
**Great job!** I really appreciate the value that this adds to the repository. The introduction of logs provides better debugging capabilities, and the peer manager successfully handles the discovery and dial of peers. 

Apart from some comments, I would also add the `peer_data.json` file to the `.gitignore`.

File: src/config.rs (hunk 1 above)
Comment (brech1) [reply src/config.rs]: And the above comment 😅

File: src/discovery.rs (hunk 2 above)
Comment (ckoopmann) [reply src/discovery.rs]: Good point. Removed the unused ones.

File: src/main.rs (hunk 4 above)
Comment (ckoopmann) [reply src/main.rs]: Good point, replaced it accordingly

File: src/main.rs (hunk 5 above)
Comment (ckoopmann) [reply src/main.rs:140]: Not sure to be honest. I just kind of randomly fiddled with these values when trying to improve peer connections. 
But I think the main difference comes from the re-dialing triggered by the peer manager. Not sure if these variables have any effect at all.

File: src/main.rs (hunk 3 above)
Comment (ckoopmann) [reply src/main.rs]: Removed them

File: src/config.rs (hunk 1 above)
Comment (ckoopmann) [reply src/config.rs]: Done

File: src/main.rs (hunk 5 above)
Comment (brech1) [reply src/main.rs:140]: I think that following this guideline from Substrate, capping the established connections per peer to `2` could be a good decision.

https://github.com/paritytech/substrate/blob/master/client/network/src/lib.rs#L288

**Pull request merged**